File: Python/17. Letter Combinations of a Phone Number.py
'''
回溯：横向遍历每个数字对应的字母，纵向递归数字
'''
class Solution:
    letterMap = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
    path = ""
    res = []
    def backtrack(self, digits, start):
        if start == len(digits):
            self.res.append(self.path)
            return
        index = int(digits[start])
        string = self.letterMap[index]
        for s in string:
            self.path += s
            self.backtrack(digits, start+1)
            self.path = self.path[:-1]
    def letterCombinations(self, digits: str) -> List[str]:
        self.res.clear()
        if not digits:
            return self.res
        self.backtrack(digits, 0)
        return self.res

File: C++/17. Letter Combinations of a Phone Number.cpp
/*
回溯：横向遍历每个数字对应的字母，纵向递归数字
*/
class Solution {
public:
    const string letterMap[10] = {
        "",
        "",
        "abc",
        "def",
        "ghi",
        "jkl",
        "mno",
        "pqrs",
        "tuv",
        "wxyz"
    };
    string path = "";
    vector<string> res;
    void backtrack(string digits, int start) {
        if (start == digits.size()) {
            res.push_back(path);
            return;
        }
        int index = digits[start] - '0';
        string s = letterMap[index];
        for (auto w : s) {
            path.push_back(w);
            backtrack(digits, start + 1);
            path.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        path.clear();
        res.clear();
        if (digits.size() == 0) return res;
        backtrack(digits, 0);
        return res;
    }
};